rfcs/0003-multi-blockchain-support.md

```diff
@@ -0,0 +1,534 @@
+      ],
+      ...
+    ]
+  },
+  
+  // Code generation for ABIs,
+  codegen: (abi) => {
+    // Return a string to be put into the output file for the ABI
+  }
+}
+```
+
+Mapping files ending with `.ts` are automatically detected wherever they appear
+in the manifest and included in builds and deployments.
+
+#### Graph TS
+
+Each chain supported in mappings comes with its own `chains/<name>.ts` file in
+the `graph-ts` repository. Each of these has at least two sections: one for
+declaring host exports that are required for mappings to interact with the
+chain, another to define chain-specific types and helpers written in
+AssemblyScript. The file `chains/ethereum.ts` file could look as follows:
+
+```js
+import { Address } from '../index'
+
+export declare namespace ethereum {
+  function call(contract: Address, fn: string, args: Array<Value>): void
+}
+
+export namespace ethereum {
+    class Value { ... }
+    class Block { ... }
+    class Transaction { ... }
+    class Event { ... }
+    class Contract { ... }
+}
+```
+
+The main `index.ts` in the `graph-ts` repository re-exports each of these chain
+integrations via
+
+```js
+export * from './chains/ethereum'
+export * from './chains/substrate'
+```
+
+So that they can be imported into subgraph mappings with
+
+```js
+import { ethereum, substrate } from './chains/substrate'
+```
+
+## Compatibility
+
+The proposed design is mostly backwards-compatible, except for a few areas:
+
+- The manifest changes require a subgraph manifest migration in Graph CLI, as
+  well as database migration for the subgraph meta data to match the new
+  structure.
+
+- Moving AssemblyScript types from e.g. `EthereumValue` to `ethereum.Value`
+  _may_ break subgraphs that directly access these types. This is ok, because
+  the majority (if not all) subgraphs only use them indirectly through generated
+  code. This does not affect subgraphs that are already deployed.
+
+## Drawbacks and Risks
+
+Generally, supporting more blockchains is an opportunity much more than it is a
+risk. It widens the reach of The Graph to communities beyond Ethereum.
+
+One risk in the design itself is that it may require a bit of experimentation to
+get the Graph Node traits for integrating arbitrary chains right. Overall, the
+work resulting from this RFC may easily cover one or two months; note that
+estimating this work is the responsibility of the engineering plan, however.
+
+## Alternatives
+
+There are nuances where the design could look different, specifically around how
+the manifest spec could be updated and whether new chain integrations are developed
+as plugins or not.
+
+## Open Questions
+
+- Is indexing from multiple chains/networks within the same subgraph something
+  we need on day one or could this be a separate RFC? My tendency is towards not
+  supporting it initially, but it may not actually be that hard.
+
+- In order for time-travel queries to work for entities, we may have to force
```

Comment:
I think we figured out that we could also segregate chains w/ different notions of causality by entity-attribute sets.